File: leetcode/TwoSum/twosum.js
// https://leetcode.com/problems/two-sum/


//Solution: Two-pass Hash Table
/*
We reduce the look up time from O(n)O(n) to O(1)O(1) by trading space for speed. A hash table 
is built exactly for this purpose, it supports fast look up in near constant time. I say "near" 
because if a collision occurred, a look up could degenerate to O(n)O(n) time. But look up in hash 
table should be amortized O(1)O(1) time as long as the hash function was chosen carefully.
*/ 

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    let map = new Map();
    const n = nums.length;
    for(let i = 0 ; i < n; i++){
        map.set(nums[i], i);
    }
    for(let i = 0; i < n; i++){
        const complement = target - nums[i];
        if(map.has(complement) && map.get(complement) != i){
            return [i, map.get(complement)];   
         }
    }
}; 

Runtime: 56 ms, faster than 85.49% of JavaScript online submissions for Two Sum.
Memory Usage: 36.3 MB, less than 11.16% of JavaScript online submissions for Two Sum.


// Solution: Two-pass Hash Table with JS Objects. 

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
     let map = new Object();
    for (var i = 0; i < nums.length; i++) {
        let complement = target - nums[i];
        if (map.hasOwnProperty(complement)) {
            return [map[complement], i]
        }
        map[nums[i]] = i;
    }
}; 

File: leetcode/1TwoSum/twosum.py
# Solution: Two pass Hash Table
# Complexity: Time O(n) | Space O(n)

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        complements = {}
        for i in range(len(nums)):
            complements[nums[i]] = i
            
        for i in range(len(nums)):
            complement = target - nums[i]
            if(complement in complements and i != complements[complement]):
                return [i, complements[complement]]
        
        return []

# Solution: One Pass Hash Table
# Complexity: Time O(n) | Space O(n)
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        complements = {}
        for i, num in enumerate(nums):
            complement = target - num
            if complement in complements:
                return [i, complements[complement]]
            complements[num] = i
        return 